Integration of api-filter-proxy with websocket-proxy

prachidamle opened the pull request:
The code changes to integratie api-filter-proxy with websocket-proxy:
- add new property to proxy config to read the file path for apifilterproxy config.json
- add a new handler to handle the requests after all other handlers have processed the requests
- If the request matches any path mentioned in apifilterproxy config.json, process the filters. Else reverseproxy the request to cattle

@cjellick please review

https://github.com/rancher/rancher/issues/7134

prachidamle commented:
@cjellick @ibuildthecloud for review/merge

alena1108 commented on proxy/apifilterproxy/filters/auth/rancher_token_validate_filter.go:
```diff
@@ -0,0 +1,277 @@
+	log.Debugf("Request => %v", input)
+
+	var cookie []string
+	if len(input.Headers["Cookie"]) >= 1 {
```
should we check for "cookie" existence here?

prachidamle replied:
I think len(input.Headers["Cookie"]) >= 1 is checking for that.

qianlifu12345 replied:
Yes, the line 62 is checking for whether the cookie existed in header.

alena1108 commented on proxy/apifilterproxy/filters/auth/rancher_token_validate_filter.go:
```diff
@@ -0,0 +1,277 @@
+	var cookieString string
+	if len(cookie) >= 1 {
+		for i := range cookie {
+			if strings.Contains(cookie[i], "token") {
```
can we check for not nil cookie[i] value before the comparison?

qianlifu12345 replied:
if cookie[i] is empty, the comparison will return false.

alena1108 commented on proxy/apifilterproxy/filters/auth/rancher_token_validate_filter.go:
```diff
@@ -0,0 +1,277 @@
+
+		}
+	} else {
+		log.Errorf("No token found")
```
is it the error situation? a bit unclear as we log in error level, but return ok

qianlifu12345 replied:
In some case , the request don't have the token. if no token found in request,return 200.

qianlifu12345 replied:
Do we need log this err in info level?

prachidamle replied:
yeah change the log level for this to debug

alena1108 commented on proxy/apifilterproxy/filters/auth/rancher_token_validate_filter.go:
```diff
@@ -0,0 +1,277 @@
+		var err error
+		if envid != "" {
+			projectID, accountID, err = getAccountAndProject(f.rancherURL, envid, tokenValue)
+			if accountID == "Unauthorized" {
```
we should check for "err" value from the line 111 before processing accountid. Is it also possible for accountId, projectId to be nil?

qianlifu12345 replied:
I will add a err check here. accountid and projectid cannot be nil. They will be empty string when it failed.

alena1108 commented on proxy/apifilterproxy/filters/auth/rancher_token_validate_filter.go:
```diff
@@ -0,0 +1,277 @@
+		if envid != "" {
+			projectID, accountID, err = getAccountAndProject(f.rancherURL, envid, tokenValue)
+			if accountID == "Unauthorized" {
+				log.Errorf("Unauthorized")
```
Would it be clear from the logs what request exactly is Forbidden/Unauthorized? As we log just exactly this word w/o any other content

qianlifu12345 replied:
Ok, I will fix this. I will add specific error log here.

prachidamle commented:
@qianlifu12345 can you address the review comments @alena1108 has for your filter?

alena1108 commented on proxy/apifilterproxy/filters/auth/rancher_token_validate_filter.go:
```diff
@@ -0,0 +1,277 @@
+				return output, nil
+			}
+			if err != nil {
+				log.Errorf("Error getting the accountid and projectid: %v", err)
```
Will it be only "not found" kind of error if the error is thrown?

alena1108 commented on proxy/apifilterproxy/filters/auth/rancher_token_validate_filter.go:
```diff
@@ -0,0 +1,277 @@
+				output.Status = http.StatusUnauthorized
+				return output, nil
+			}
+			if err != nil {
```
same here. It would probably make sense to check for not nil error first

alena1108 commented on proxy/apifilterproxy/filters/auth/rancher_token_validate_filter.go:
```diff
@@ -0,0 +1,277 @@
+		return "Forbidden", "Forbidden", err
+
+	}
+	if projectid == userid {
```
Could you briefly explain why projectId being equal to userId, is an indication of project id validation failure?

alena1108 commented on proxy/apifilterproxy/filters/auth/rancher_token_validate_filter.go:
```diff
@@ -0,0 +1,277 @@
+//get the accountID from rancher API
+func getAccountID(host string, token string) (string, error) {
+
+	client := &http.Client{}
```
can we create a common http client for all the requests executed against Rancher? And then just reference it from all the methods that need it.

alena1108 commented on proxy/apifilterproxy/filters/auth/rancher_token_validate_filter.go:
```diff
@@ -0,0 +1,277 @@
+	cookie := http.Cookie{Name: "token", Value: token}
+	req.AddCookie(&cookie)
+	resp, err := client.Do(req)
+	if err != nil {
```
Looks like validation done at lines 176-198 and 227-249 is quite similar. Is it possible to incorporate it to some centralized place?

alena1108 commented on proxy/apifilterproxy/filters/auth/rancher_token_validate_filter.go:
```diff
@@ -0,0 +1,277 @@
+
+		idData, suc := messageData.Data[i].(map[string]interface{})
+		if suc {
+			id, suc := idData["id"].(string)
```
we should check for "id" presence in idData first before casting it to string

alena1108 commented on proxy/apifilterproxy/filters/auth/rancher_token_validate_filter.go:
```diff
@@ -0,0 +1,277 @@
+		idData, suc := messageData.Data[i].(map[string]interface{})
+		if suc {
+			id, suc := idData["id"].(string)
+			kind, namesuc := idData["kind"].(string)
```
same here. Check if present before casting

alena1108 commented on proxy/apifilterproxy/handler.go:
```diff
@@ -0,0 +1,182 @@
+	if err != nil {
+		//failed to reload the config from the config.json
+		log.Debugf("Reload failed with error %v", err)
+		returnHTTPError(w, r, http.StatusInternalServerError, "Failed to reload the proxy config")
```
shall we report the exact error here? Given it is getting logged in Debug, user wont have content of what has happened

alena1108 commented on proxy/apifilterproxy/handler.go:
```diff
@@ -0,0 +1,182 @@
+
+	destProxy, err := newProxy(destination)
+	if err != nil {
+		log.Errorf("Error creating a reverse proxy for destination %v", destination)
```
We should log the exact error here. Otherwise user won't know what went wrong

alena1108 commented on proxy/apifilterproxy/manager/filter_manager.go:
```diff
@@ -0,0 +1,200 @@
+
+func ProcessPreFilters(path string, api string, body map[string]interface{}, headers map[string][]string) (map[string]interface{}, map[string][]string, string, model.ProxyError) {
+	prefilters := PathPreFilters[path]
+	log.Debugf("START -- Processing pre filters for request path %v", path)
```
we should probably make the level at least configurable. So if user want all the requests to be in, he can configure it.

prachidamle replied:
Unfortunately we don't have a mechanism to change the log level for microservices launched by cattle within the container.

alena1108 commented on proxy/apifilterproxy/util/util.go:
```diff
@@ -0,0 +1,29 @@
+	return encodedSignature
+}
+
+func GenerateUUID() string {
```
we have to document on how we generate the UUID so the people writing generic http proxy implementation know how to parse it

alena1108 commented on proxy/apifilterproxy/manager/filter_manager.go:
```diff
@@ -0,0 +1,200 @@
+	refChan := make(chan int, 1)
+	refreshReqChannel = &refChan
+
+	if configFile != "" {
```
I think we already check for configFile not being empty on line 50

alena1108 commented on proxy/apifilterproxy/filters/auth/rancher_token_validate_filter.go:
```diff
@@ -0,0 +1,277 @@
+		return output, nil
+	}
+	if tokenValue == "" {
+		log.Errorf("No token found")
```
same here. If status is ok, log in debug

alena1108 commented on proxy/apifilterproxy/filters/auth/rancher_token_validate_filter.go:
```diff
@@ -0,0 +1,277 @@
+	if tokenValue == "" {
+		log.Errorf("No token found")
+		output.Status = http.StatusOK
+		return output, nil
```
do we ever return not nil error in this function? If not - and the only thing we do is create ok vs non ok output status - may be it would make sense to remove the error from the list of returned arguments?

alena1108 commented on proxy/apifilterproxy/filters/auth/rancher_token_validate_filter.go:
```diff
@@ -0,0 +1,277 @@
+	req.AddCookie(&cookie)
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Errorf("Cannot connect to the rancher server. Please check the rancher server URL")
```
we already log the error in the caller stack. There is no need to log it here. What you can do is - return the error in the format:

```
return ("", "", fmt.Errorf("Something failed due to [%v]", err))
```

Same for all the code blocks below

qianlifu12345 committed:
fix bugs in code review

prachidamle committed:
Code review changes

prachidamle committed:
Trash and Vendor changes for api-filter-proxy integration

prachidamle committed:
Code changes for Integration of api-filter-proxy with websocket-proxy

cjellick commented on proxy/apifilterproxy/filters/http/generic_http_filter.go:
```diff
@@ -0,0 +1,75 @@
+		return output, err
+	}
+
+	log.Debugf("Request => " + string(bodyContent))
```
the correct way to write this would be log.Debugf("Request => %s", bodyContent)
that way, the type conversion (which involves a potentially expensive copy) only occurs when debug is on (bc the params are passed to Debugf which checks to see if debug is on before doing the conversion)

prachidamle replied:
Updated this.

cjellick commented on proxy/apifilterproxy/model/filter_model.go:
```diff
@@ -0,0 +1,24 @@
+package model
+
+import ()
```
No need for an empty import, right?

prachidamle replied:
Removed the empty import

cjellick commented on proxy/apifilterproxy/filters/auth/rancher_token_validate_filter.go:
```diff
@@ -0,0 +1,289 @@
+	Data []interface{} `json:"data,omitempty"`
+}
+
+func init() {
```
init() functions are only called when a package is imported. As far as I can tell, these filter packages (http and autn) are not imported anywhere. How does this work?

cjellick replied:
Nevermind. I found them.

prachidamle replied:
yeah they are "imported" in manager/filter_manager.go using "_"

cjellick replied:
Using init() this way is a bad pattern as it couples the filters to the manager while hiding the relationship in an anonymous import. 

Can you eliminate the init functions? I think you could incorporate this change with refactoring the manager package into a struct.

prachidamle replied:
We wanted to have a "pluggable" filter model where people can write against the api_filter interface and just drop in the go file. They would specify any parameters they need to initialize their filters in env and load them in init()
That is why we are using this model (similar to the external-dns/external-lb pluggable frameworks)

prachidamle replied:
How do you want to refactor this into manager package? Can you write up a small pseudo code for me to understand?

cjellick replied:
It isn't any more pluggable by using an init() function. You still have to load it from another package. Each handler package can still have NewHandler() function that does the same loading from environment variables. The change will just be to call NewHandler explicitly instead of doing `import _ github.com/blah/blah`

cjellick replied:
Quick pseudo code, your mileage may very:
```
func InitHandler(filterConfigFile string, cattleAddr string) (*APIFiltersHandler, bool) {
    filters := getFilters() 
    manager.SetConfig(filterConfigFile, cattleAddr)
    router := newFilterRouter(manager.ConfigFields, filters)
    
    filterHandler = &APIFiltersHandler{filterRouter: router}
    return filterHandler, manager.APIFilterProxyReady
}

func getFilters() map[string]handler {
   m // handler map
  authHandler := auth.NewHandler()
  m[authHandler.getName()] = authhandler
  
  // repeat for other handlers.
}
```
I realize this seems like you are creating a dependency between this package and each of the filters, but you already had the dependency, you were just obfuscating it through the anonymous import

cjellick replied:
I also think you can combine the logic in the manager package with the logic in the filters/api_filter.go and apfilterproxy/handler.go

You should be able to achieve one cohesive struct that encapsulates all this logic. The manager package should not have to expose any package level variables publicly.

prachidamle replied:
Refactored the code as per above.

cjellick commented on proxy/apifilterproxy/manager/filter_manager.go:
```diff
@@ -0,0 +1,199 @@
+	"github.com/rancher/websocket-proxy/proxy/apifilterproxy/util"
+)
+
+var (
```
We should get away from using package level variables like this. Can you refactor this into a struct with properties?

prachidamle replied:
removed the package level vars and refactored the code

cjellick commented on proxy/proxy.go:
```diff
@@ -235,13 +240,19 @@ func newCattleProxies(config *Config) (*proxyProtocolConverter, *cattleWSProxy)
 func (h *proxyProtocolConverter) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 	proxyprotocol.AddHeaders(req, h.httpsPorts)
-	h.reverseProxy.ServeHTTP(rw, req)
+	if h.apiFilterProxyReady {
```
Is this suppose to reflect the value of `manager.APIFilterProxyReady`? because the value is copied upon returning from `apifilterproxy.InitHandler()`, so changes to `manager.APIFilterProxyReady` will not be reflected here.

cjellick replied:
On that note, I don't think you should have this switch logic here. the cattleProxy should be a property on your proxy and in your proxy, if no path is matched, you should called the cattleProxy

prachidamle replied:
Yes, I was trying to play safe while integrating this to fallback on your existing cattleproxy. But I dont think I need it anymore, will always call my proxy to handle.

prachidamle replied:
Removed this boolean

cjellick commented on proxy/apifilterproxy/manager/filter_manager.go:
```diff
@@ -0,0 +1,199 @@
+func Reload() error {
+	//put msg on channel, so that any other request can wait
+	select {
+	case *refreshReqChannel <- 1:
```
Don't use a channel here. Use a RWMutex. All writes to properties should obtain a write lock and then all reads (meaning the http handlers that use the properties) should obtain a read lock.

prachidamle replied:
Removed the channel and added RWMutex

cjellick commented:
I don't want to merge this without tests.

prachidamle committed:
Refactored the code:
1) Addressed code review comments about api_filters init() way of
plugging the filters
2) Added the manager code under a cohesive struct
3) Added RWmutex to lock/unlock read and writes to filter config fields
4) WSP forwards the requests through apiproxy always
